### ISSUE TITLE
Challenge: ee.jakarta.tck.json.bind.defaultmapping.polymorphictypes.AnnotationTypeInfoTest

### ISSUE THREAD

**Issue by jeanouii** [jakartaee/jsonb-api#347]
Per the current [TCK Process documentation](https://jakarta.ee/committees/specification/tckprocess/), please find bellow a challenge related to JSON-B TCK. Can someone please add the `challenge` label which is required by the TCK Process but not available to everyone to add.

### The relevant specification version and section number(s)

https://jakarta.ee/specifications/jsonb/3.0/jakarta-jsonb-spec-3.0.html#custom-instantiation

### The coordinates of the challenged test(s)

Only following test in the class are related to the challenge 

* #testCreatorDeserialization

### The exact TCK version

JSON-B TCK version 3.0.0

### The implementation being tested, including name and company

Apache Johnzon 2.0.0-SNAPSHOT from the Apache Software Foundation

### A full description of why the test is invalid and what the correct behavior is believed to be

The test aims at covering the polymorphic mapping using JsonbTypeInfo and JsonbSubtype which is fine. The DateConstructor, in its current form 

```
    public static final class DateConstructor implements SomeDateType {

        public LocalDate localDate;

        @JsonbCreator
        public DateConstructor(@JsonbProperty("localDate") @JsonbDateFormat(value = "dd-MM-yyyy", locale = "nl-NL") LocalDate localDate) {
            this.localDate = localDate;
        }
    }
```

There is an ambiguity there in the way the localDate field should be deserialized. We have 2 candidates
- the factory (constructor from DateConstruct)
- the setter (direct field access which is public without a private setter)

The test implies that only one, the DateConstructor constructor and not the direct field.

In Johnzon we call first the factory and then the setters, because we see them as callback for the user to use after the object is constructed.

In this situation, the specification does not provide with a clear order. I'm going to create a issue to clarify this point but meanwhile, we need a challenge for this test.

I propose to create a PR in order to change the localDate modifier from public to private so there is no ambiguty anymore. We can add a public getter so we can even write an assert in the test using the deserialized value.

**Comment by jungm** [comment]
Any chance we can get this challenge accepted and resolved? Its one of the two excludes we carry on right now in apache johnzon that blocks us from actually being certified

Sorry for the ping @KyleAure but i've seen you active here the last couple weeks and maybe we can finally get this done

**Comment by KyleAure** [comment]
@jeanouii @jungm 

I agree that there is some ambiguity in the specification that needs to be handled, but I'm not sure the way Johnzon is handling this situation is how I would like the specification settle on.

If I understand correctly the issue here is name collision.
For example if we have the following POJO: 

```java
public class CreatorNameConflicts {
    public String fromJson;
    public String fromCreator; 
    public String notProvided; 
    public int notDeclared;

    @JsonbCreator
    public CreatorNameConflicts(
          @JsonbProperty("fromJson") String fromJson,
          @JsonbProperty("notDeclared") int notDeclared) {
        this.fromCreator = fromJson;
        this.notDeclared = notDeclared;
    }
}
```

And we perform the following deserialization: 

```java
CreatorNameConflicts result = jsonb.fromJson("{\"fromJson\":\"jsonValue\"}", CreatorNameConflicts.class);
```

Then the way you have described the behavior in Johnzon, deserialization would result in the following object: 

```txt
CreatorNameConflicts [ fromJson: "jsonValue", fromCreator: "jsonValue", notProvided: null, notDeclared: 0 ]
```

I think this goes against the intention of the `@JsonbCreator` to create an `alternative` path for deserialization. 
The spec says: 

```txt
In many scenarios instantiation with the use of default constructor is not enough. To support these scenarios, JSON Binding provides jakarta.json.bind.annotation.JsonbCreator annotation.
```

I think the intention here is to say that when using the `@JsonbCreator` we do not intend to go down the default deserialization path.

Not to mention if you were to switch the `CreatorNameConflicts` class with an equivalent record then `fromJson` would have to be `null` since what is returned from the constructor would be immutable.

I would like our behavior between classes and records to be as closely aligned as possible.

I tried to clarify this in this PR: https://github.com/jakartaee/jsonb-api/pull/411
I would appreciate your feedback on this PR.

**Comment by KyleAure** [comment]
For the purpose of 3.0, I do agree the behavior is ambiguous and am willing to accept this challenge and disable the test in a 3.0.3 service release to be used by Johnzon for certification, but I think for 3.1.0 I'd rather clarify the ambiguous behavior and leave the test as is.

I'd appreciate input for others as well.

**Comment by rmannibucau** [comment]
it is important to keep setters or field initialization cause they are different from constructor/factory initialization in terms of callback, lifecycle and backward compatibility - worse for records than pojo, ack that.

so don't think we can say we do bypass them, in particular

the immutable case is the extreme constructor case where there is only ambiguities for public fields (let say never else you dont use a jsonbcreator and just let it be bound by the public fields) but all in between cases require both support so (obviously) I think Apache Johnzon is the only viable option else you enforce to move all the logic to the constructor which is not always possible or factories with potentially a lot of duplication when you mix programming model on your DTO (you can say it is wrong but it is not that uncommon and vert convenient for migrations so think we should cover these case)

hope it makes sense (from the phone)

**Comment by KyleAure** [comment]
@rmannibucau 

I agree that public fields and setters should still be used when avaiablle, just not when they have conflicting names with parameters in the `@JsonbCreator`. 

For example if we deserialized the following: 

```java
CreatorNameConflicts result = jsonb.fromJson("{\"fromJson\":\"jsonValue\", \"notProvided\": \"jsonValue2"}", CreatorNameConflicts.class);
```

I would expect the following instance to be returned: 

```txt
CreatorNameConflicts [ fromJson: null, fromCreator: "jsonValue", notProvided: "jsonValue2", notDeclared: 0 ]
```

~~Which sadly I do not think we currently do in Yasson.~~

Does that align with your comment?

**Comment by jungm** [comment]
> I think the intention here is to say that when using the @JsonbCreator we do not intend to go down the default deserialization path.

Disagree with that one, IMO its just about how the object is being created. Other properties should still be mapped, this behaviour is also required by the TCK already in `testCustomConstructorPlusFields`.

I do however agree with Kyles standpoint that once a property has been passed to a creator it should be seen as "consumed" and not be passed on to mapping it into setters/fields anymore for a future spec release

**Comment by rmannibucau** [comment]
Why not allowing to consume twice since the lifecycle is clearly defined by design there?

While I get the "we can prevent multiple mapping" - a set more or less technically, I dont see why making it an exception would help, if done it is intentional from user standpoint so we should respect it and not try to be clever IMHO.

Exceptions are always overcomplex for a poor gain, fine to mention implémentation can log a warning or a like.
Ideally I would be happy to enable to "eat" attributes explicitly (@JsonbIgnoreDeserialization - with a better name) either on the prop or paramétrer, but not the opposite.

**Comment by KyleAure** [comment]
@jungm 
> > I think the intention here is to say that when using the @JsonbCreator we do not intend to go down the default deserialization path.

> Disagree with that one, IMO its just about how the object is being created. Other properties should still be mapped, this behaviour is also required by the TCK already in testCustomConstructorPlusFields.

When I said `we do not intend to go down the default deserialization path` my intention was to say that the purpose of `@JsonbCreator` is itself to define an alternative path for deserialization.  Therefore a user declaring a constructor with `@JsonbCreator` would expect that alternative path to be used. 

However, the way you have described Johnzon's implementation there would be scenarios where it *looks* like that alternative path is skipped.  Namely, when there are name conflicts on between parameters on the `@JsonbCreator` constructor/method and the fields of the class / bean properties. 

@rmannibucau 
> While I get the "we can prevent multiple mapping" - a set more or less technically, I dont see why making it an exception would help, if done it is intentional from user standpoint so we should respect it and not try to be clever IMHO.

In a sense I agree, but IMO if a user were to write the following: 

```java
public class Test {
  public String a;
  @JsonbCreator
  public Test(@JsonbProperty("a") String a) {
    this.a = a.toUpperCase();
  }
```

Then I think the intention is clear that they want `a` to be uppercase at the end of deserialization.
However, IIUC you are saying that the users intention is the opposite because they didn't set the field `a` on the class `Test` to a different name?

**Comment by rmannibucau** [comment]
> Then I think the intention is clear that they want a to be uppercase at the end of deserialization.

Can agree on this one, will disagree if you update your sample to use `public void setA(String v)` and I don't think making public fields an exception would be very understandable and neat for end users, therefore my proposal to just stay dumb and use all options which are available. Similarly factories are different from constructors (cause "external" of the bean self lifecycle) but I dont think making constructor an exception (compared to static methods) would be understandable (+ valuable).

So overall I think it makes sense to not try to be clever but maybe only document these cases in the spec?

**Comment by KyleAure** [comment]
> So overall I think it makes sense to not try to be clever but maybe only document these cases in the spec?

Sounds good. I'll draft a PR and see if we can come to a consensus on language.
